File: samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/api/ApiServerApplication.java
package systems.zlink.samples.tictactoe.server.api;

import org.slf4j.LoggerFactory;
import org.springframework.boot.WebApplicationType;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.builder.SpringApplicationBuilder;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.boot.web.context.WebServerInitializedEvent;
import org.springframework.boot.web.server.WebServerFactoryCustomizer;
import org.springframework.boot.web.servlet.server.ConfigurableServletWebServerFactory;
import org.springframework.context.ApplicationListener;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.context.annotation.Bean;
import org.springframework.core.env.StandardEnvironment;

import systems.zlink.framework.locations.redis.ZLinkRedisLocationStore;
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime;
import systems.zlink.framework.spring.EnableZLinkFramework;
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer;
import systems.zlink.samples.tictactoe.server.api.handlers.AuthenticatePlayerHandler;
import systems.zlink.samples.tictactoe.server.api.handlers.CreateGameHttpHandler;
import systems.zlink.samples.tictactoe.server.configuration.ApiSettings;
import systems.zlink.samples.tictactoe.server.configuration.SampleLocationStore;
import systems.zlink.samples.tictactoe.server.configuration.TicTacToeReadinessReporter;

import java.net.InetAddress;
import java.net.URI;
import java.nio.file.Path;

@EnableZLinkFramework
@EnableConfigurationProperties(ApiSettings.class)
@SpringBootApplication(
        proxyBeanMethods = false,
        scanBasePackageClasses = {
            ApiServer.class,
            AuthenticatePlayerHandler.class,
            CreateGameHttpHandler.class
        })
public final class ApiServerApplication {
    private ApiServerApplication() {}

    public static ConfigurableApplicationContext run(String configPath) {
        StandardEnvironment environment = new StandardEnvironment();
        environment
                .getPropertySources()
                .remove(StandardEnvironment.SYSTEM_ENVIRONMENT_PROPERTY_SOURCE_NAME);
        environment
                .getPropertySources()
                .remove(StandardEnvironment.SYSTEM_PROPERTIES_PROPERTY_SOURCE_NAME);
        SpringApplicationBuilder builder =
                new SpringApplicationBuilder(ApiServerApplication.class)
                        .environment(environment)
                        .web(WebApplicationType.SERVLET)
                        .properties(
                                "spring.config.location="
                                        + Path.of(configPath).toAbsolutePath().toUri());
        builder.application().setKeepAlive(true);
        return builder.run(new String[0]);
    }

    @Bean
    ZLinkFrameworkConfigurer apiFramework(ApiSettings settings) {
        return ApiServer.configure(settings);
    }

    @Bean(destroyMethod = "close")
    ZLinkRedisLocationStore locationStore(ApiSettings settings) {
        return SampleLocationStore.create(settings);
    }

    @Bean
    WebServerFactoryCustomizer<ConfigurableServletWebServerFactory> apiHttpServer(
            ApiSettings settings) {
        return server -> {
            var endpoint = URI.create(settings.apiBindUrl());
            server.setAddress(InetAddress.getLoopbackAddress());
            server.setPort(endpoint.getPort());
        };
    }

    @Bean
    ApplicationListener<WebServerInitializedEvent> ticTacToeHttpReadiness(ApiSettings settings) {
        return event ->
                LoggerFactory.getLogger(ApiServerApplication.class)
                        .info("tictactoe-ready kind=http node={}", settings.nodeId());
    }

    @Bean(destroyMethod = "close")
    TicTacToeReadinessReporter ticTacToeSpotRouteReadiness(
            ApiSettings settings, ZLinkRouteMeshRuntime meshes) {
        return TicTacToeReadinessReporter.spotRoute(settings.nodeId(), meshes);
    }
}

File: samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/api/ApiServerApplication.kt
package systems.zlink.samples.kotlin.tictactoe.server.api

import java.net.InetAddress
import java.net.URI
import java.nio.file.Path
import org.slf4j.LoggerFactory
import org.springframework.boot.WebApplicationType
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.builder.SpringApplicationBuilder
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.boot.web.context.WebServerInitializedEvent
import org.springframework.boot.web.server.WebServerFactoryCustomizer
import org.springframework.boot.web.servlet.server.ConfigurableServletWebServerFactory
import org.springframework.context.ApplicationListener
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.context.annotation.Bean
import org.springframework.core.env.StandardEnvironment
import systems.zlink.framework.locations.redis.ZLinkRedisLocationStore
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime
import systems.zlink.framework.spring.EnableZLinkFramework
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer
import systems.zlink.samples.kotlin.tictactoe.server.api.handlers.AuthenticatePlayerHandler
import systems.zlink.samples.kotlin.tictactoe.server.api.handlers.CreateGameHttpHandler
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleLocationStore
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleSettings
import systems.zlink.samples.kotlin.tictactoe.server.configuration.TicTacToeReadinessReporter

@EnableZLinkFramework
@EnableConfigurationProperties(SampleSettings::class)
@SpringBootApplication(
    proxyBeanMethods = false,
    scanBasePackageClasses =
        [ApiServer::class, AuthenticatePlayerHandler::class, CreateGameHttpHandler::class],
)
class ApiServerApplication {
    @Bean
    fun apiFramework(settings: SampleSettings): ZLinkFrameworkConfigurer =
        ApiServer.configure(settings)

    @Bean(destroyMethod = "close")
    fun locationStore(settings: SampleSettings): ZLinkRedisLocationStore =
        SampleLocationStore.create(settings)

    @Bean(destroyMethod = "close")
    fun ticTacToeReadinessReporter(
        settings: SampleSettings,
        meshes: ZLinkRouteMeshRuntime,
    ): TicTacToeReadinessReporter = TicTacToeReadinessReporter.api(settings.nodeId, meshes)

    companion object {
        fun run(configPath: String): ConfigurableApplicationContext {
            val environment =
                StandardEnvironment().apply {
                    propertySources.remove(
                        StandardEnvironment.SYSTEM_ENVIRONMENT_PROPERTY_SOURCE_NAME
                    )
                    propertySources.remove(
                        StandardEnvironment.SYSTEM_PROPERTIES_PROPERTY_SOURCE_NAME
                    )
                }
            return SpringApplicationBuilder(ApiServerApplication::class.java)
                .also { builder -> builder.application().setKeepAlive(true) }
                .environment(environment)
                .web(WebApplicationType.SERVLET)
                .properties(
                    "spring.config.location=${Path.of(configPath).toAbsolutePath().toUri()}"
                )
                .run()
        }
    }

    @Bean
    fun apiHttpServer(
        settings: SampleSettings
    ): WebServerFactoryCustomizer<ConfigurableServletWebServerFactory> =
        WebServerFactoryCustomizer { server ->
            val endpoint = URI.create(settings.apiBindUrl)
            server.setAddress(InetAddress.getLoopbackAddress())
            server.setPort(endpoint.port)
        }

    @Bean
    fun ticTacToeHttpReadiness(
        settings: SampleSettings
    ): ApplicationListener<WebServerInitializedEvent> = ApplicationListener {
        LoggerFactory.getLogger(ApiServerApplication::class.java)
            .info("tictactoe-ready kind=http node={}", settings.nodeId)
    }
}
